accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -495,10 +498,26 @@ of a function call produces a result of type `Never`.
 
 The static type of a `throw e` expression is `Never`.
 
-In legacy mode, an override of operator== with no explicit parameter type
+An expression `e` of the form `e1 == e2` is treated as
+`let v1 = e1, v2 = e2 in v1 != null ? (v2 != null ? v1.==(v2) : false) : (v2 !=
```

Comment:
I don't think this de-sugaring captures what we want - as written, it implies that:

```
int x;
FutureOr<int?> y;
x == y;
```
is an error.  You clarify further down in the commentary, but I think the result is confusing.  I'd prefer to specify this directly - I think it's a lot clearer.  That is, say that:

Given `e1 == e2` of static type `T1` and `T2` respectively:
  -  Let `S` be the declared type of the parameter to the `operator ==` method on `T1` as defined elsewhere
  - It is an error if `T1` is not assignable to `S?`.
    - Note specifically that it is not necessarily an error if `T1` is not assignable to `S`.

Reply:
Amazingly, given a promotable variable `v`, `v != null ? ../* v isn't null */.. : ....` will obviously make the claim in the comment sound, but since null promotion relies completely on **NonNull**, it is actually not always known that `v` is non-null at that point.

The test you mention will drop a little bit of type information, though:

```dart
class A {
  operator ==(covariant A other) => ...;
  ...
}

void main() {
  A a = A();
  A? aq = a;
  a == 1; // Compile-time error: `1` does not have type `A`.
  aq == 2; // OK.
}
```

We _do_ know that also the second test will fail (based on the statically known type of the receiver in the situation where the `operator ==` is actually invoked), but the parameter type will be `Object` because the type of `aq` is nullable. So we could rely on 'Let `S` be the declared type of the parameter to the `operator ==` method on **NonNull**(`T1`)' to include this extra information, at least for the cases where **NonNull** is able to eliminate the null.

Reply:
Resolving -- this was rewritten completely.